Add support for new parameters

11harveyj opened the pull request:
Opening this PR to add support for new parameters and functions that were not in previous versions.

Majority of video functions, amp functions, all tone functions and listening modes will be added. Below is a check list that I need to work through to get all the responses decoded, then I will work through this list again to implement any functions that currently are not implemented. This is all taken from official Pioneer documentation

I have _loads_ more changes for ha-pioneer_async lined up to include support for all of this.

- [x] Power
- [x] Volume
- [x] Mute
- [x] Input
- [x] Tone Control
- [x] Listening Modes
- [x] DSP
- [x] Channel Levels
- [x] AMP
- [x] Key Lock
- [x] Video
- [x] Zone Power
- [x] Zone Input
- [x] Zone Volume
- [x] Zone Mute
- [x] Zone Channel Level
- [x] Zone Tone
- [x] Information Requests
- [x] Tuner
- [x] iPod Operation
- [x] NETWORK Operation
- [x] ADAPTER PORT Operation
- [x] MHL Device Operation
- [x] Cursor Operation

crowbarz commented:
Looking and sounding great. I got the basic functionality (on/off, input switching, volume) that I needed working and unfortunately everything else went on the backlog. The next items on my list was decoding the audio and video parameters to be able to detect stereo vs multichannel audio, and being able to tune the radio frequency - both of which are already on your list :)

11harveyj commented:
Thanks, the biggest one for me was the listening modes, the remote for my AVR doesn't work very well anymore (despite replacing batteries) and the iControl app has completely stopped working on my phone, while I was at it I thought to myself "well, some other features in the app are quite useful" so decided to just go through all the docs and serial commands / responses.

I've just finished audio / video information decoding but I need some help as at the moment it is only able to if we sent ?AST or ?VST, which I need to send every time we receive input changes or power state changes.

Other solution to this (maybe) is that at least on my AVR I get a response starting with AUA / AUB when the source changes but I've found no documentation on what that actually is yet.

crowbarz commented:
Sending another command when the parser (run from the connection listener task) detects a state change is a little tricky - you need to schedule a separate task to call `send_command()` to avoid blocking the connection listener task. This is currently done in `_connection_listener()` to schedule `bounce_volume()` when AVR power-on is detected if `PARAM_POWER_ON_VOLUME_BOUNCE` is enabled, you should be able to do something similar when an input or power state change is detected.

11harveyj commented:
Using inspiration from the `bounce_volume()` functions and how that's scheduled in, I've added a command queue, which can be used to schedule in commands to run after processing the response by adding `commands_to_queue.add("COMMAND HERE")` in the `_parse_response()` function.

I don't think `AST` / `VST` will be sent in the serial session if that state changes on the AVR itself (IE, HDMI input goes from stereo to multi channel etc.)

Tuner functions complete, a few actions I haven't added as other functions take care of what they would do, list below:

- TUNER PRESET (DIGIT key) - `set_tuner_preset()` takes care of this, but combines the class and preset digit together
- TUNER CLASS change - Same as above
- TUNER PRESET INCREMENT - Same as above
- TUNER PRESET DECREMENT - Same as above
- DIRECT ACCESS (tuner) - This sets frequency directly, but responses are more complicated to implement, for now the loop in `set_tuner_frequency()` takes care of this

No specific iPod functionality added, apart from just mapping the commands and being able to send those commands, might be useful in HA where we can have play/pause buttons etc. I'm not able to test this though as I don't own an iPod.

11harveyj commented:
I think this is ready now, there a probably some places that could be improved on, happy to change where needed.

These two commands have been added now:

TUNER PRESET INCREMENT
TUNER PRESET DECREMENT

They are used as previous and next commands for the media controls...

I need a place to document the media controls though, would you be happy to enable to wiki feature in this repo and that could then be used to store the Python API docs perhaps? Currently it doesn't return playback information, however I do have the docs for that but I'm not sure if the AVRs will automatically send playback info when that state changes or if I'll have to schedule more frequent updates for those.

crowbarz commented:
I enabled the wiki, but I think you need to first have submitted to this repo in order for you to update the wiki.

I started reviewing the PR, it works with my VSX-930 so that's a great start :) I have a few initial general comments and will add additional comments to the code as I go (may take a while, there's a lot to digest):

* There's an unfortunate clash of terminology here as I've used "parameter" to refer to operational attributes that could differ between different AVR models, such as maximum possible volume or a flag for a feature that exists only for some AVR models. But I note that Pioneer also uses "parameter" to refer to command and response arguments (as well as audio/video parameter!). I guess this might be the rationale for the naming of `PARAM_MEDIA_CONTROL_SOURCES`, `PARAM_MEDIA_CONTROL_COMMANDS`, etc? To avoid conflation between these two uses, I'd suggest just dropping the `PARAM_` prefix for these, they seem to me to just be dicts used to translate response values to something more meaningful.
* It would be good for those dicts to be moved to a separate file, eg. `const.py`. This could be a good future home for `PIONEER_COMMANDS` to reduce `pioneer_avr.py` by ~700 lines, down from 3200+ lines (a bit long I think!) though let's leave that change for a separate PR.
* A full update takes about 5 seconds on my AVR as it queries a lot of attributes now. It may be worth gating the updates behind parameters for some high level categories (eg.  `tone`, `amp`, `tuner`, `channel_level`, `dsp`, `video`, `audio`) that can be turned off by default and turned on only for specific models via model specific default parameters or user level parameter overrides. This would minimise backwards compatibility and reduce impact to AVRs that may not have such features or be able to handle all the new queries.
* These groups would probably also be useful when exposing the attributes to HA, as I'd ideally prefer to avoid having to update the HA integration whenever more attributes are added.
* note to self: will need to refactor `bounce_task` to use the command queue, on my AVR that is running in parallel to the full update that is triggered at AVR power on.

I've also updated `cli.py` to add a few commands that dump the various new attributes, and will submit that to your fork shortly.

11harveyj commented:
> There's an unfortunate clash of terminology here as I've used "parameter" to refer to operational attributes that could differ between different AVR models, such as maximum possible volume or a flag for a feature that exists only for some AVR models. But I note that Pioneer also uses "parameter" to refer to command and response arguments (as well as audio/video parameter!). I guess this might be the rationale for the naming of PARAM_MEDIA_CONTROL_SOURCES, PARAM_MEDIA_CONTROL_COMMANDS, etc? 
> To avoid conflation between these two uses, I'd suggest just dropping the PARAM_ prefix for these, they seem to me to just be dicts used to translate response values to something more meaningful.
It would be good for those dicts to be moved to a separate file, eg. const.py. This could be a good future home for PIONEER_COMMANDS to reduce pioneer_avr.py by ~700 lines, down from 3200+ lines (a bit long I think!) though let's leave that change for a separate PR.

Agree with all of this, I've updated now and dropped PARAM, we can move a lot out to a `const.py` file as suggested later on.

> A full update takes about 5 seconds on my AVR as it queries a lot of attributes now. It may be worth gating the updates behind parameters for some high level categories (eg. tone, amp, tuner, channel_level, dsp, video, audio) that can be turned off by default and turned on only for specific models via model specific default parameters or user level parameter overrides. This would minimise backwards compatibility and reduce impact to AVRs that may not have such features or be able to handle all the new queries.
> These groups would probably also be useful when exposing the attributes to HA, as I'd ideally prefer to avoid having to update the HA integration whenever more attributes are added.

Noted, I've updated now and added the following user configurable parameters:

`PARAM_DISABLE_AUTO_QUERY` - disables all queries when `_update_zone` is called.
`PARAM_ENABLED_FUNCTIONS` - provides the ability to only query certain "functions", in future we can use this against some of the set commands, but for now it splits the `PIONEER_COMMAND` key on an underscore, and checks if the second item in that array is in `PARAM_ENABLED_FUNCTIONS`.

Currently the following is supported for `PARAM_ENABLED_FUNCTIONS`:

- amp
- dsp
- tuner
- tone
- channels
- video
- system
- audio

crowbarz commented on aiopioneer/param.py:
```diff
@@ -24,12 +33,180 @@
+    PARAM_ZONE_3_SOURCES: ["04", "06", "15", "26", "38", "53", "41", "44", "45", "17", "13", "05", "01", "02", "33", "46", "47", "99", "10"],
+    PARAM_HDZONE_SOURCES: ["25", "04", "06", "10", "15", "19", "20", "21", "22", "23", "24", "34", "35", "26", "38", "53", "41", "44", "45", "17", "13", "33", "31", "46", "47", "48"],
+    PARAM_HDZONE_VOLUME_REQUIREMENTS: ["13", "15", "05", "25"],
+    PARAM_SPEAKER_SYSTEM_MODES: {
```
Are the `PARAM_SPEAKER_SYSTEM_MODES`, `PARAM_LISTENING_MODES` and `PARAM_VIDEO_RESOLUTION_MODES` text description mappings the same across AVR models, or do they differ across AVR models? If the mappings are the same, then can we keep only the valid mode IDs in the params rather than the entire mapping, and move the mapping (containing all possible values) to a const.

11harveyj replied:
So originally those were static, however as I was developing I found that IDs and names map differently across different generations of AVRs, for example, some of the mappings of those on my SC-LX57 (FY14 docs on the Pioneer website) were different when compared to newer generations of AVRs. Hence why I've put them where it is, I'd like to clean it up a bit, but I'm not sure the best way to do that, given each generation seems to have a slightly different set of values.

Saying that `PARAM_LISTENING_MODES` I _think_ is generally static across the generations (just different models support different modes), so that would certainly shorten the code by using a list of IDs rather than a dict.

11harveyj replied:
Updated and changed the behaviour of `PARAM_LISTENING_MODES` and `PARAM_VIDEO_RESOLUTION_MODES`.

Rather than a list of enabled listening modes, I've gone for a list of disabled listening modes.

I've left `PARAM_SPEAKER_SYSTEM_MODES` as it was because those values change depending on AVR and generation of AVR.

crowbarz commented on aiopioneer/pioneer_avr.py:
```diff
@@ -1125,6 +2814,30 @@ async def bounce_volume(self):
+    async def _command_queue_schedule(self, command_queue):
+        """Schedule commands to queue."""
+        _LOGGER.debug(">> PioneerAVR._command_queue_schedule()")
+        await self._command_queue_cancel()
```
It looks to me that the command queue subtask is spawned with the commands passed to the subtask via arguments. There's a problem here that `_command_queue_schedule()` first cancels any currently running command queue task before scheduling a new task with a new set of commands, so the set of commands being executed by the running task is lost when it is cancelled.

I'd suggest instead queuing the commands in `self._command_queue`, this is already defined in `__init__()` but doesn't seem to be used. So the logic for the scheduler should be something like:

1. Check whether the task is already created
2. If so, add the additional commands in `self._command_queue`
3. If not, create a new task that starts going through the command queue

You will probably need to change the iteration in `_execute_command_queue` to use a `while` loop, as funny things happen when you iterate through lists that mutate during the `for` loop. Also, if a particular query command is already in the command queue, then there should probably be no need to queue it again.

crowbarz commented on aiopioneer/pioneer_avr.py:
```diff
@@ -1191,3 +2903,303 @@ async def mute_off(self, zone="1"):
+            ## Set the tuner band
+            await self.send_command("set_tuner_band_fm", zone, ignore_error=False)
+
+        ## Round the frequency to nearest 0.05 if band is FM, otherwise divide frequency by 9 using modf so that the remainder is split out, then select the whole number response and times by 9
```
Be aware that AM frequencies are 10kHz apart in the Americas ... so you might not want to round.
I thought I previously saw a command for inputting numbers to set a frequency? That would be more reliable than stepping up and down, though the command might not be supported on all AVR models.

11harveyj replied:
I think there is a system setting to configure the step in AM frequencies so I'll use that.

you are correct, there is a function to set a specific frequency, but it would be a lot more _complicated_ to implement than other commands, and direct access isn't support on all AVR models, whereas increasing and decreasing the tuner frequency is supported on all models (unless of course the tuner does not exist). I went for the most supportable approach that works regardless of AVR generation.

crowbarz commented on aiopioneer/pioneer_avr.py:
```diff
@@ -1191,3 +2903,303 @@ async def mute_off(self, zone="1"):
+            to_freq = (str(frequency))
+            current_freq = (str(self.tuner.get("frequency")))
+
+            if to_freq == current_freq:
```
If the AVR doesn't step to the exact requested frequency, then this loop will probably go forever. So probably want the loop to stop also when you have gone past the requested frequency.

crowbarz commented:
I made some changes to the command queue functions to minimise repeated commands being sent to the AVR, and also migrated the volume bouncer to use the command queue instead, can you see if either of these impact your changes.

I haven't tested the new `set_*` functions yet, they can be tested later. VScode/pylint is also complaining about a few other issues but I think these can also be addressed later on.

The above notwithstanding I think this PR is pretty much ready to merge, after which I can push a beta release to PyPi and test further within my prod environment, and also work on the HA integration to expose the additional AVR info and functions to HA. Thanks again @11harveyj for your efforts

11harveyj commented:
Ok, great. I'll test these changes later on and feedback :)

11harveyj commented:
Everything is still working ok for me, my HA instance is pending an update so when I update that I'll test those changes in HA too (changed the manifest.json to point to my fork).
